synapse/storage/prepare_database.py

```diff
@@ -278,20 +278,31 @@ def _upgrade_existing_database(
             the current_version wasn't generated by applying those delta files.
         database_engine (DatabaseEngine)
         config (synapse.config.homeserver.HomeServerConfig|None):
-            application config, or None if we are connecting to an existing
-            database which we expect to be configured already
+            None if we are initialising a blank database, otherwise the application
+            config
         data_stores (list[str]): The names of the data stores to instantiate
             on the given database.
         is_empty (bool): Is this a blank database? I.e. do we need to run the
             upgrade portions of the delta scripts.
     """
+    if is_empty:
+        assert not applied_delta_files
+    else:
+        assert config
```

Comment:
Not sure if it matters (or if using assertions like this is common in Synapse), but people like to run things with `-O` which strips them...

Reply:
my logic in using `assert` is that we'll (probably) only get here in the case of some sort of programming error elsewhere, in which case the assertions will get picked up during testing, so it would be ok if the `asserts` got stripped when run in production.

And yes, there are a few other assertions around to check for programming errors like this.

I don't feel super-strongly about it either way.

Reply:
Makes sense. Thanks for the explanation!